src/daft-io/src/object_io.rs

```diff
@@ -112,10 +117,203 @@ pub(crate) trait ObjectSource: Sync + Send {
     }
 }
 
+pub(crate) async fn glob(
+    source: Arc<dyn ObjectSource>,
+    glob: &str,
+) -> super::Result<BoxStream<'static, super::Result<FileMetadata>>> {
+    let glob_fragments = to_glob_fragments(glob)?;
+    let full_glob_matcher = GlobBuilder::new(glob)
+        .literal_separator(true)
+        .backslash_escape(true)
+        .build()
+        .map_err(|err| super::Error::InvalidArgument {
+            msg: format!("Cannot parse provided glob {glob}: {err}"),
+        })?
+        .compile_matcher();
+
+    // Channel to send results back to caller. Note that all results must have FileType::File.
+    let (to_rtn_tx, mut to_rtn_rx) = tokio::sync::mpsc::channel(16 * 1024);
+
+    /// Dispatches a task to visit the specified `path` (a concrete path on the filesystem to either a File or Directory).
+    /// Based on the current glob_fragment being processed (accessible via `glob_fragments[i]`) this task will:
+    ///   1. Perform work to retrieve Files/Directories at (`path` + `glob_fragments[i]`)
+    ///   2. Return results to the provided `result_tx` channel based on the provided glob, if appropriate
+    ///   3. Dispatch additional tasks via `.visit()` to continue visiting them, if appropriate
+    fn visit(
+        result_tx: Sender<super::Result<FileMetadata>>,
+        source: Arc<dyn ObjectSource>,
+        path: &str,
+        glob_fragments: (Arc<Vec<GlobFragment>>, usize),
+        full_glob_matcher: Arc<GlobMatcher>,
+    ) {
+        let path = path.to_string();
+        tokio::spawn(async move {
+            log::debug!(target: "glob", "Visiting '{path}' with glob_fragments: {glob_fragments:?}");
+            let (glob_fragments, i) = glob_fragments;
+            let current_fragment = &glob_fragments[i];
+
+            // BASE CASE: current_fragment is a **
+            // We perform a recursive ls and filter on the results for only FileType::File results that match the full glob
+            if current_fragment.escaped_str() == "**" {
+                let next_level_file_metadata =
+                    source.iter_dir(path.as_str(), Some("/"), None).await;
+
+                match next_level_file_metadata {
```

Comment:
you can factor this into a closure and use the `?` operator inside instead that and push the result of the function into the channel

Reply:
Could you elaborate or provide a short code snippet?

Reply:
I'm still not quite sure which lines of code you're suggesting to move into a closure, but take a look at this change and see if this makes it cleaner?

https://github.com/Eventual-Inc/Daft/pull/1450/commits/6f78a112405fa3af5f37b9709e556d4692947e82